doc/programming/guidelines.rst

```diff
@@ -0,0 +1,291 @@
+little additional complexity to the code. On top of that, overall
+performance often depends on how features are combined, or layered
+across several units.
+
+Hence, it's useful to remember the following four rules on Spicy
+performance:
+
+    1. Do as much work as needed in your Spicy analyzer, but not more.
+       For example, what you can do in Zeek, do there.
+
+    2. If you're in doubt about the performance impact of some Spicy
+       code, do a benchmark.
+
+    3. Rule 2: If you think you know the performance impact of some
+       Spicy code, do a benchmark!
+
+    4. In practice, it might all not matter that much anyways!
+
+To explain Rule 4: When you are writing a parser for a network
+protocol in Spicy, the resulting performance is a direct factor of the
+amount of traffic that parser will be processing. For example, in a
+typical Zeek setup, where you network carries a mix of various
+protocols, your parser will see only a subset of the overall traffic.
+And if you aren't going for the handful of most common high-volume
+protocols (e.g., HTTP, DNS, TLS), then your parser will very likely
+end up processing only a tiny fraction of the overall traffic. At that
+point, its runtime performance is going to be dwarfed by everything
+else Zeek is doing. Hence, take the following with a grain of salt.
+Usually it's best to get you parser working first, then benchmark to
+see if you need to improve performance, and finally find the primary
+bottlenecks of needed.
+
+.. _performance_runtime:
+
+Runtime Performance
+-------------------
+
+Among other things, runtime performance is affected by:
+    - additional C++ code that is generated by the Spicy compiler
+    - use of expensive Spicy constructs
+    - poorly managed objects hogging memory (both on stack and heap)
+
+Below are some strategies to make the runtime performance as efficient as possible.
+
+.. rubric:: Avoid declaring public units
+
+The Spicy compiler generates additional code for ``public`` units,
+thereby increasing both compile and execution times. Defining units as
+``public`` is only required for top-level units where parsing starts
+(i.e., with Zeek: the units that you enter into your EVT file). The
+attribute can be omitted otherwise.
+
+.. rubric:: Turn unreferenced vectors into anonymous fields
+
+Implement vectors that don't need to be referenced from your code as
+:ref:`anonymous fields <anonymous_fields>` (i.e., do not provide a
+name for the field). Normal, named fields store the whole vector
+accumulating all parsed elements over the lifetime of the defining
+unit; whereas anonymous vector field forgo storage, because nobody
+could access the fields anyways.
+
+The most common pitfall here is top-level units that parse a sequence
+of PDUs:
+
+.. code::
+
+    public type MyPDUs = unit {
+        msgs: Message[]; # accumulates all messages until end of session
+        : Message[];     # anonymous field, no accumulation
```

Comment:
I would clearly call out antipatterns, e.g., here

```suggestion
        msgs: Message[]; # DANGEROUS: accumulates all messages until end of session
        : Message[];     # SUGGESTED: anonymous field, no accumulation
```

There might be more instances below where this might help people stay clear of copying the incorrect line.